pkg/policies/controlplane/components/gradient-controller.go

```diff
@@ -0,0 +1,142 @@
+		return nil, nil, fmt.Errorf("max_gradient must be greater than min_gradient")
+	}
+
+	gradient := &GradientController{
+		slope:             gradientParameters.Slope,
+		minGradient:       gradientParameters.MinGradient,
+		maxGradient:       gradientParameters.MaxGradient,
+		componentID:       componentID.String(),
+		dynamicConfigKey:  gradientControllerProto.ManualModeConfigKey,
+		defaultManualMode: gradientControllerProto.ManualMode,
+		manualMode:        gradientControllerProto.ManualMode,
+	}
+
+	return gradient, fx.Options(), nil
+}
+
+// Name implements runtime.Component.
+func (g *GradientController) Name() string {
+	return "Gradient"
+}
+
+// Type implements runtime.Component.
+func (g *GradientController) Type() runtime.ComponentType {
+	return runtime.ComponentTypeSignalProcessor
+}
+
+// ShortDescription implements runtime.Component.
+func (g *GradientController) ShortDescription() string {
+	return "Gradient Controller"
+}
+
+// IsActuator implements runtime.Component.
+func (g *GradientController) IsActuator() bool {
+	return false
+}
+
+// Execute implements runtime.Component.Execute.
+func (g *GradientController) Execute(inPortReadings runtime.PortToReading, circuitAPI runtime.CircuitAPI) (outPortReadings runtime.PortToReading, err error) {
+	signal := inPortReadings.ReadSingleReadingPort("signal")
+	setpoint := inPortReadings.ReadSingleReadingPort("setpoint")
+	optimize := inPortReadings.ReadSingleReadingPort("optimize")
+	max := inPortReadings.ReadSingleReadingPort("max")
+	min := inPortReadings.ReadSingleReadingPort("min")
+	controlVariable := inPortReadings.ReadSingleReadingPort("control_variable")
+	var output runtime.Reading
+
+	// ComputeOutput
+	if setpoint.Valid() && signal.Valid() && controlVariable.Valid() {
+		var gradient float64
+		gradient = math.Pow(signal.Value()/setpoint.Value(), g.slope)
+		// clamp to min/max
+		if gradient < g.minGradient {
+			gradient = g.minGradient
+		}
+		if gradient > g.maxGradient {
+			gradient = g.maxGradient
+		}
+		if math.IsNaN(gradient) {
+			output = runtime.InvalidReading()
+		} else {
+			output = runtime.NewReading(controlVariable.Value() * gradient)
+		}
+	} else {
+		output = runtime.InvalidReading()
+	}
+
+	// Optimize
+	if output.Valid() && optimize.Valid() {
+		targetOutput := runtime.NewReading(output.Value() + optimize.Value())
+		output = targetOutput
+	}
+
+	if output.Valid() {
+		outputReading := output
+		// Constrain output
+		if max.Valid() {
+			outputReading = runtime.NewReading(math.Min(output.Value(), max.Value()))
+		}
+		if min.Valid() {
+			outputReading = runtime.NewReading(math.Max(outputReading.Value(), min.Value()))
+		}
+		if outputReading.Value() != output.Value() {
+			output = outputReading
+		}
+	}
+
+	// Set output to control variable in-case of Manual mode
+	if g.manualMode {
+		output = controlVariable
+	}
+
+	// Save readings for the next tick so that Controller may access them via ControllerStateReadAPI
+	g.lastOutput = output
```

Comment:
No need to remember the last reading

Reply:
[63c267f](https://github.com/fluxninja/aperture/pull/2838/commits/63c267ff138b9e85f4f18813a91aa8bf78e58e8e) addresses this.